webpush/vapid.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webpush
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"math/big"
+	"net/url"
+	"time"
+)
+
+var (
+	// JWT header is always the same, so use a pre-encoded string
+	vapidPrefix = []byte("eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.")
+	dot         = []byte(".")
+)
+
+type jwtBody struct {
+	Aud string `json:"aud"`
+	Sub string `json:"sub,omitempty"`
+	Exp int64  `json:"exp"`
+}
+
+// Vapid provides helpers for authenticating the sender of a push message
+type Vapid struct {
+	// The application's private key
+	priv *ecdsa.PrivateKey
+
+	// Sub should be an email or URL, for identification
+	Sub string
+}
+
+// PublicKey returns an uncompressed EC point representing the
+// application's public key
+func (v *Vapid) PublicKey() []byte {
+	pub := v.priv.PublicKey
+	return elliptic.Marshal(pub.Curve, pub.X, pub.Y)
+}
+
+// Token creates a token with the specified endpoint, using configured Sub id
+// and a default expiration (1h).
+func (v *Vapid) Token(aud string) (string, error) {
```

Comment:
"... with the specified endpoint **aud**, using configured **subscription ID** ..."
expiration is hard-coded so it is not a default. just say something like "Token is valid for the next hour".